dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+
+    public async Task SignalAsync(string signal, object[] args, RPCOptions? options = null) { /*...*/ }
+
+    public async Task SignalAsync(Func<Task> signal, RPCOptions? options = null) { /*...*/ }
+
+    public async Task SignalAsync<T>(Func<T, Task> signal, T arg, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<TQueryResult>(
+      string query, object[] args, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<TQueryResult>(
+      Func<TQueryResult> query, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<T, TQueryResult>(
+      Func<T, TQueryResult> query, T arg, RPCOptions? options = null) { /*...*/ }
+
+    public async Task CancelAsync(RPCOptions? options = null) { /*...*/ }
+
+    public async Task TerminateAsync(
+        string reason? = null, object[]? args = null, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<WorkflowExecutionDescription> DescribeAsync(RPCOptions? options = null) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public async Task<WorkflowHistory> FetchHistoryAsync(
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+    public IAsyncEnumerator<HistoryEvent> FetchHistoryEvents(
+        bool waitNewEvent = false,
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+#endif
+}
+```
+
+Notes:
+
+* Notice you can provide a return type explicitly on the `GetResultAsync` call if the generic one on the handle isn't
+  what you want
+* Like start workflow, signal/query string-based calls accept an array of objects instead of trying to use varargs or
+  separating single arg vs multi. This makes the overload count better and will give users compile time errors if it's
+  not literally an object array
+* Notice that `TerminateAsync` does not _require_ `args` like start, signal, and query calls that accept `args` arrays
+  because we don't think it's important for users to explicitly send empty if they don't want them like we think they
+  should for the others
+* Same issue with `FetchHistory*` as `ListWorkflows` on client - no `IAsyncEnumerator` on really old .NET versions, but
+  we're ok with that
+
+## Worker
+
+Usage:
+
+```csharp
+using Temporal.Client;
+using Temporal.Worker;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Build worker options
+var options = new TemporalWorkerOptions("task queue");
+options.AddActivities(new GreetingActivities());
+
+// Run until Ctrl+C
+var worker = new TemporalWorker(client, options);
+var cts = new CancellationTokenSource();
+Console.CancelKeyPress += (sender, eventArgs) => {
```

Comment:
Ok I have to say this is the nicest way of adding an additional signal handler I've seen in a language for a minute